Library/Homebrew/utils.rb

```diff
@@ -770,3 +770,30 @@ def number_readable(number)
   (numstr.size - 3).step(1, -3) { |i| numstr.insert(i, ",") }
   numstr
 end
+
+# True if this version of Ruby supports text encodings in its strings
+def ruby_has_encoding?
+  String.method_defined?(:force_encoding)
+end
+
+# Truncates a text string to fit within a byte size constraint,
+# preserving character encoding validity. The returned string will
+# be not much longer than the specified max_bytes, though the exact
+# shortfall or overrun may vary.
+def truncate_text_to_approximate_size(s, max_bytes, options = {})
+  front_weight = options.fetch(:front_weight, 0.5)
+  return s if s.bytesize <= max_bytes
+
+  if front_weight < 0.0 || front_weight > 1.0
+    raise "opts[:front_weight] must be between 0.0 and 1.0"
+  end
+  glue = "\n[...snip...]\n"
+  max_bytes_in = [max_bytes - glue.bytesize, 1].max
+  bytes = ruby_has_encoding? ? s.force_encoding("BINARY") : s
+  glue_bytes = ruby_has_encoding? ? glue.encode("BINARY") : glue
+  front = bytes[0..(max_bytes_in * front_weight).floor]
+  back = bytes[-(max_bytes_in * (1.0 - front_weight)).ceil..-1]
+  out = front + glue_bytes + back
+  out.force_encoding("UTF-8").encode("UTF-16", :invalid => :replace).encode!("UTF-8") if ruby_has_encoding?
```

Comment:
Sorry for being a bit annoying, but you missed to change the other `encode` to `encode!`. This line is pretty long and convoluted. How about rewriting this as follows?

```ruby
out = front + glue_bytes + back
if ruby_has_encoding?
  out.force_encoding("UTF-8")
  out.encode!("UTF-16", :invalid => :replace)
  out.encode!("UTF-8")
end
out
```